xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,470 @@
+  }
+
+  @VisibleForTesting
+  final class WeightedRoundRobinPicker extends ReadyPicker {
+    private final List<Subchannel> list;
+    private final AtomicReference<EdfScheduler> schedulerRef;
+    private volatile boolean rrMode = true;
+
+    WeightedRoundRobinPicker(List<Subchannel> list, int startIndex) {
+      super(checkNotNull(list, "list"), startIndex);
+      Preconditions.checkArgument(!list.isEmpty(), "empty list");
+      this.list = list;
+      this.schedulerRef = new AtomicReference<>(new EdfScheduler(list.size()));
+      updateWeight();
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      if (rrMode) {
+        return super.pickSubchannel(args);
+      }
+      int pickIndex = schedulerRef.get().pick();
+      WrrSubchannel subchannel = (WrrSubchannel) list.get(pickIndex);
+      if (!config.enableOobLoadReport) {
+        return PickResult.withSubchannel(
+           subchannel,
+           OrcaPerRequestUtil.getInstance().newOrcaClientStreamTracerFactory(
+               subchannel.perRpcListener));
+      } else {
+        return PickResult.withSubchannel(subchannel);
+      }
+    }
+
+    private void updateWeight() {
+      EdfScheduler scheduler = new EdfScheduler(list.size());
+      int weightedChannelCount = 0;
+      double avgWeight = 0;
+      for (Subchannel value : list) {
+        double newWeight = ((WrrSubchannel) value).getWeight();
+        if (newWeight > 0) {
+          avgWeight += newWeight;
+          weightedChannelCount++;
+        }
+      }
+      if (weightedChannelCount < 2) {
+        rrMode = true;
+        return;
+      }
+      avgWeight /= 1.0 * weightedChannelCount;
+      for (int i = 0; i < list.size(); i++) {
+        WrrSubchannel subchannel = (WrrSubchannel) list.get(i);
+        double newWeight = subchannel.getWeight();
+        scheduler.add(i, newWeight > 0 ? newWeight : avgWeight);
+      }
+      schedulerRef.set(scheduler);
+      rrMode = false;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(WeightedRoundRobinPicker.class)
+          .add("list", list).add("rrMode", rrMode).toString();
+    }
+
+    @VisibleForTesting
+    List<Subchannel> getList() {
+      return list;
+    }
+
+    @Override
+    public boolean isEquivalentTo(RoundRobinPicker picker) {
+      if (!(picker instanceof WeightedRoundRobinPicker)) {
+        return false;
+      }
+      WeightedRoundRobinPicker other = (WeightedRoundRobinPicker) picker;
+      // the lists cannot contain duplicate subchannels
+      return other == this
+          || (list.size() == other.list.size() && new HashSet<>(list).containsAll(other.list));
+    }
+  }
+
+  /**
+   * The earliest deadline first implementation in which each object is
+   * chosen deterministically and periodically with frequency proportional to its weight.
+   *
+   * <p>Specifically, each object added to chooser is given a deadline equal to the multiplicative
```

Comment:
I'd strongly discourage equating "deadline" with "1/weight". I'd suggest something more like "the deadline advances by the multiplicative inverse of the object's weight." (And just remove the 1/weight two lines down)